python/lsst/ts/dream/csc/dream_csc.py

```diff
@@ -479,6 +525,140 @@ async def send_events(self, status_data: dict[str, Any]) -> None:
+                    )
+
+            # Second attempt: Fresh request and save locally
+            async with client.stream("GET", dream_url) as response:
+                if response.status_code != 200:
+                    self.log.error(
+                        f"Data lost [2]: failed to retrieve file: {dream_url} - HTTP {response.status_code}"
+                    )
+                    return
+
+                try:
+                    await self.save_to_local_disk(response, key)
+                except Exception:
+                    self.log.exception("Also failed to save file locally. Data lost!")
+
+    async def save_to_s3(self, response: httpx.Response, key: str) -> None:
+        """Upload file to S3 from an HTTP stream.
+
+        Parameters
+        ----------
+        response: httpx.Response
+            HTTP response object for the file to save.
+
+        key: str
+            S3 style filename key to save to.
+        """
+        if not self.s3bucket:
+            raise RuntimeError("S3 bucket not configured")
+
+        try:
+            with io.BytesIO(await response.aread()) as buffer:
+                await self.s3bucket.upload(fileobj=buffer, key=key)
+            url = (
+                f"{self.s3bucket.service_resource.meta.client.meta.endpoint_url}/"
+                f"{self.s3bucket.name}/{key}"
+            )
+            await self.evt_largeFileObjectAvailable.set_write(
+                url=url,
+                generator="dream",
+            )
+            self.log.info(f"Successfully uploaded {key} to S3.")
+        except Exception:
+            self.log.exception(f"Failed to upload {key} to S3.")
+            raise
+
+    async def save_to_local_disk(self, response: httpx.Response, key: str) -> None:
+        """Save the file from an HTTP stream to local disk.
+
+        Parameters
+        ----------
+        response: httpx.Response
+            HTTP response object for the file to save.
+
+        key: str
+            S3 style filename key to save to.
+        """
+        if not self.s3bucket:
+            raise RuntimeError("S3 bucket not configured")
+
+        try:
+            filepath = pathlib.Path("/tmp") / self.s3bucket.name / key
+            dirpath = filepath.parent
+            if not dirpath.exists():
+                self.log.info(f"Creating directory {str(dirpath)}")
+                dirpath.mkdir(parents=True, exist_ok=True)
+
+            with open(filepath, "wb") as tmp_file:
+                async for chunk in response.aiter_bytes():
+                    tmp_file.write(chunk)
+
+            await self.evt_largeFileObjectAvailable.set_write(
+                url=filepath.as_uri(),
+                generator="dream",
+            )
+            self.log.info(f"Saved {key} to local disk at {filepath}")
+        except Exception:
+            self.log.exception("Could not save the file to local disk.")
+            raise
```

Comment:
Same as above. Consider just removing the try/except clause here and handle the exception at another layer.